tracer/src/Datadog.Trace/AppSec/BodyExtractor.cs

```diff
@@ -0,0 +1,189 @@
+// <copyright file="BodyExtractor.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Datadog.Trace.AppSec.Waf;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.AppSec
+{
+    internal static class BodyExtractor
+    {
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor(typeof(BodyExtractor));
+
+        private static readonly HashSet<Type> AdditionalPrimitives = new()
+        {
+            typeof(string),
+            typeof(decimal),
+            typeof(Guid),
+            typeof(DateTime),
+            typeof(DateTimeOffset),
+            typeof(TimeSpan)
+        };
+
+        internal static object Extract(object body)
+        {
+            var visted = new HashSet<object>();
+            var item = ExtractType(body.GetType(), body, 0, visted);
+
+            return item;
+        }
+
+        private static bool IsOurKindOfPrimitive(Type t)
+        {
+            return t.IsPrimitive || AdditionalPrimitives.Contains(t) || t.IsEnum;
+        }
+
+        private static Dictionary<string, object> ExtractProperties(object body, int depth, HashSet<object> visited)
+        {
+            if (visited.Contains(body))
+            {
+                return new Dictionary<string, object>(0);
+            }
+
+            visited.Add(body);
+
+            Log.Debug("ExtractProperties - body: {Body}", body);
+
+            var fields = body.GetType()
+                        .GetFields(BindingFlags.Instance | BindingFlags.NonPublic)
```

Comment:
Reflection and LINQ combo! How often does this code execute? If we visit the same `body` type over and over, should we think of caching the `fields` array per type?

Could we at least remove the LINQ bits and do the equivalent filtering in the `for` loop below? If nothing else, it will at least avoid several heap allocations.